FrEee/Scripts/AI/Vanguard/Plan.csx

```diff
@@ -0,0 +1,278 @@
+        CurrentPlans = new PlanList();
+        Empire.LogAIMessage("Failed to load plans"); 
+        
+    }
+
+    public static void Pack()
+    {
+        CurrentPlans.AssignAllIds(); 
+
+        if (Empire.AINotes.ContainsKey("PlanList"))
+            Empire.AINotes.Remove("PlanList");
+
+        var serializedPlans = FrEee.Utility.JsonSerializer.SerializeObjectToJson(CurrentPlans);
+        Empire.LogAIMessage(serializedPlans); 
+        Empire.AINotes.Add("PlanList", serializedPlans); 
+    }
+
+
+    public static void RemoveCompleted()
+    {
+        CurrentPlans.ColonizationPlans.RemoveAll(x => x.IsComplete);
+        CurrentPlans.ShipBuildPlans.RemoveAll(x => x.IsComplete); 
+    }
+
+    
+}
+
+/// <summary>
+/// The method of storing plans between turns for the AI. 
+/// </summary>
+[Serializable]
+public class PlanList
+{
+    public List<ColonizationPlan> ColonizationPlans;
+
+    public List<ShipBuildPlan> ShipBuildPlans; 
+
+    [DoNotSerialize]
+    public IEnumerable<Plan> AllPlans { 
+        get
+        {
+            var plans = new List<Plan>();
+            plans.AddRange(ColonizationPlans);
+            plans.AddRange(ShipBuildPlans);
+            return plans; 
+        } 
+    }
+
+
+    public PlanList()
+    {
+        ColonizationPlans = new List<ColonizationPlan>();
+        ShipBuildPlans = new List<ShipBuildPlan>(); 
+    }
+
+    public void AssignAllIds()
+    {
+        foreach (var p in ColonizationPlans)
+            if (p.PlanId == null)
+                p.AssignId();
+        foreach (var p in ShipBuildPlans)
+            if (p.PlanId == null)
+                p.AssignId();
+    }
+
+    /// <summary>
+    /// gets the chosen plan, or null. 
+    /// </summary>
+    /// <param name="planId"></param>
+    /// <returns></returns>
+    public Plan GetPlan(string planId)
+    {
+        return AllPlans.FirstOrDefault(x => x.PlanId == planId); 
+    }
+}
+
+
+[Serializable]
+public abstract class Plan
+{
+    public string PlanId; 
+
+
+    [DoNotSerialize]
+    public virtual string TypeShorthand { get; }
+
+    public bool IsComplete;
+
+    /// <summary>
+    /// The priority of the plan. Higher is better.  
+    /// </summary>
+    public int Priority; 
+
+
+    public void AssignId()
+    {
+        PlanId = $"{TypeShorthand}-{DateTime.UtcNow.Ticks}-{PlanManager.CurrentPlans.AllPlans.Count()}";
+    }
+}
+
+
+[Serializable]
+public class ColonizationPlan : Plan
+{
+    [DoNotSerialize]
+    public override string TypeShorthand => "C";
+
+    /// <summary>
+    /// The planet intended to colonize. 
+    /// </summary>
+    [DoNotSerialize]
+    public Planet Planet { get {
+            if (PlanetId == 0)
+                return null; return GalaxyReference<Planet>.GetGalaxyReference(PlanetId);  } set { PlanetId = value.ID; } }
+
+    /// <summary>
+    /// The planet that will build the colony ship to colonize this world. 
+    /// </summary>
+    [DoNotSerialize]
+    public Planet SourcePlanet { get
```

Comment:
What if we want to build a colony ship at a base or mobile space yard, or use an existing colony ship?

Reply:
At present, a preexisting colony ship can be assigned to the plan. Though will fail to load if not at a colony. 

However, for simplicity, a colony ship can only be built at planets to make sure there is population to load. Might tweak that to allow orbital yards when they are implemented.

Reply:
Makes sense!